### PR TITLE
Add sample that demonstrates how to use HP Tuning with containers on …

### PR DESCRIPTION
**Pull request by nnegrey**
…ML Engine

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/cloudml-samples/359)

### CODE REVIEW

File: pytorch/containers/hp_tuning/task.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                        type=float,
+                        default=0.5,
+                        help='SGD momentum (default: 0.5)')
+    parser.add_argument('--seed',
+                        type=int,
+                        default=42,
+                        help='random seed (default: 42)')
+    args = parser.parse_args()
+
+    task(args)
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/task.py]: nit: function name should be more descriptive (Example: train_model)
http://google.github.io/styleguide/pyguide.html

File: pytorch/containers/hp_tuning/task.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        if args.job_dir:
+            # Save the model to GCS
+            data_utils.save_model(args.job_dir, args.model_name)
+    else:
+        print('Accuracy: {:.0f}%'.format(latest_accuracy))
+
+
+def main():
+    # Training settings
+    parser = argparse.ArgumentParser(description='PyTorch Sonar Example')
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/task.py]: nit: create a different method for reading args, as other samples: (Example: get_args()) https://github.com/GoogleCloudPlatform/cloudml-samples/blob/master/boston/tensorflow/keras/trainer/task.py#L28

File: pytorch/containers/hp_tuning/task.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    if item < 0.5:
+        return 0.0
+    else:
+        return 1.0
+
+
+def test(net, test_loader):
+    """Test the DNN"""
+    net.eval()
+    criterion = nn.BCELoss()
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/task.py]: nit: Can we add some details on the implementation of nn.BCEloss(). Maybe a link/paper

File: pytorch/containers/hp_tuning/task.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    correct = 0
+
+    with torch.no_grad():
+        for i, data in enumerate(test_loader, 0):
+            features = data['features']
+            target = data['target']
+            output = net(features)
+            # Binarize the output
+            pred = output.apply_(set_value_)
+
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/task.py]: nit: extra space

File: pytorch/containers/hp_tuning/data_utils.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        shuffle=True)
+    test_loader = DataLoader(
+        test_dataset.dataset,
+        batch_size=batch_size,
+        shuffle=True)
+
+    return train_loader, test_loader
+
+
+def save_model(job_dir, model_name):
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/data_utils.py]: nit: Maybe in function documentation add a summary of changes for clarity?

File: pytorch/containers/hp_tuning/task.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    test_loss = 0
+    correct = 0
+
+    with torch.no_grad():
+        for i, data in enumerate(test_loader, 0):
+            features = data['features']
+            target = data['target']
+            output = net(features)
+            # Binarize the output
+            pred = output.apply_(set_value_)
```
Comment (gogasca) [inline pytorch/containers/hp_tuning/task.py]: nit: lambda may reduce the need of set_value options [optional]

File: pytorch/containers/hp_tuning/README.md [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+# hptuning_config.yaml
+trainingInput:
+  scaleTier: BASIC
+  masterConfig:
+    imageUri: $IMAGE_URI
+  hyperparameters:
+    goal: MAXIMIZE
+    maxTrials: 10
+    maxParallelTrials: 5
+    hyperparameterMetricTag: my_metric_tag
```
Comment (dizcology) [inline pytorch/containers/hp_tuning/README.md]: if the metric used in the training code is something real, perhaps use a better name such as `my_accuracy_tag`.

File: pytorch/containers/hp_tuning/README.md [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+
+# Part 5: Deploy the docker image to Cloud Container Registry
+You should have configured docker to use Cloud Container Registry, found
+[here](https://cloud.google.com/container-registry/docs/pushing-and-pulling).
+```
+docker push $IMAGE_URI
+```
+
+# Part 6: Submit your training job
+Creates a hptuning config file for your job request.
```
Comment (dizcology) [inline pytorch/containers/hp_tuning/README.md]: add link to docs explaining how hptuning on CMLE works.

File: pytorch/containers/hp_tuning/README.md (hunk 7 above)
Comment (nnegrey) [reply pytorch/containers/hp_tuning/README.md]: Will update, what bugs me though is that on the console, that tag isn't even used. They just have: `objectiveValue`

File: pytorch/containers/hp_tuning/README.md (hunk 8 above)
Comment (nnegrey) [reply pytorch/containers/hp_tuning/README.md]: I have up at the top an overview of HP Tuning links under pre-req.
